File: typescript/trigridscaffolding.ts
// To get ts types, run: npm i --save-dev @types/readline

namespace TriGridScaffolding {
    let width: number = 120
    const DELAY: number = 60;
    let CHANGE_AMOUNT: number = 0.04;

    
    async function sleep(milliseconds: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, milliseconds));
    }

    process.on('SIGINT', function() {
        console.log("\nTri Grid Scaffolding by Al Sweigart al@inventwithpython.com");
        process.exit();
    });

    function getTerminalWidth(): number {
        if (process.stdout.isTTY) {
            return process.stdout.columns - 1; // -1 is because Windows puts a newline when printing to the last column
        } else {
            return 80;
        }
    }

    async function main(): Promise<void> {
        let density = 0.0;
        while (true) {
            width = getTerminalWidth();
            let triangleWidth: number = Math.floor((width - 2) / 4);
            let row1: string = '';
            let row2: string = '';

            density += CHANGE_AMOUNT;
            if (density < 0.0 || density >= 1.0) {
                CHANGE_AMOUNT *= -1;
            }
        
            for (let j = 0; j < 2; j++) {
                if (j === 0) {
                    // On j == 0, handle the two rows of begins-with-rightside-up-triangles:
                    //  /\  /\  /\
                    // /__\/__\/__\
                    row1 = '';
                    row2 = '';
                } else if (j == 1) {
                    // On j == 1, handle the two rows of begins-with-upside-down-triangles:
                    // \  /\  /
                    // _\/__\/_
                    row1 = '\\ ';
                    row2 = '_\\';
                }
                
                for (let i = 0; i < triangleWidth; i++) {
                    if (Math.random() < density) {
                        row1 += ' /';
                        row2 += '/';
                    } else {
                        row1 += '  ';
                        row2 += ' ';
                    }

                    if (Math.random() < density) {
                        row2 += '__';
                    } else {
                        row2 += '  ';
                    }

                    if (Math.random() < density) {
                        row1 += '\\ ';
                        row2 += '\\';
                    } else {
                        row1 += '  ';
                        row2 += ' ';
                    }
                }

                console.log(row1);
                await sleep(DELAY);
                console.log(row2);
                await sleep(DELAY);
            }
        }
    }

    main();
}

File: python/trigridscaffolding.py
import random, time, os

r"""
 /\  /\  /\  /\  /\  /\  /\  /\  /\  /\
/__\/__\/__\/__\/__\/__\/__\/__\/__\/__\
\  /\  /\  /\  /\  /\  /\  /\  /\  /\  /
_\/__\/__\/__\/__\/__\/__\/__\/__\/__\/_
"""

DELAY = 0.1

def main():
    changeAmt = 4
    density = 0

    while True:
        # -1 because Windows adds newlines if anything
        # is printed in the rightmost column.
        width = os.get_terminal_size()[0] - 1
        triangleWidth = (width - 2) // 4

        # Increase/decrease the density of triangles, until you
        # reach 0 or 100, then reverse the direction of density change:
        density += changeAmt
        if density <= 0 or density >= 100:
            changeAmt *= -1

        for j in range(2):
            # On j == 0, handle the two rows of begins-with-rightside-up-triangles:
            #  /\  /\  /\
            # /__\/__\/__\
            if j == 0:
                row1 = []
                row2 = []
            # On j == 1, handle the two rows of begins-with-upside-down-triangles:
            # \  /\  /
            # _\/__\/_
            elif j == 1:
                row1 = ['\\ ']
                row2 = ['_\\']
            for i in range(triangleWidth):
                if random.randint(0, 99) < density:
                    row1.append(' /')
                    row2.append('/')
                else:
                    row1.append('  ')
                    row2.append(' ')

                if random.randint(0, 99) < density:
                    row2.append('__')
                else:
                    row2.append('  ')

                if random.randint(0, 99) < density:
                    row1.append('\\ ')
                    row2.append('\\')
                else:
                    row1.append('  ')
                    row2.append(' ')
            print(''.join(row1)); time.sleep(DELAY)
            print(''.join(row2)); time.sleep(DELAY)


try:
    main()
except KeyboardInterrupt:
    print('Tri Grid Scaffolding, by Al Sweigart al@inventwithpython.com')



